blockchain/upgrade.go

```diff
@@ -1789,6 +1789,1162 @@ func upgradeToVersion8(ctx context.Context, db database.DB, dbInfo *databaseInfo
+				unspentBits := oldSerialized[offset]
+				for j := uint32(0); j < 8; j++ {
+					if unspentBits&0x01 != 0 {
+						// The first 2 outputs are encoded via the
+						// header code, so adjust the output number
+						// accordingly.
+						outputNum := 2 + i*8 + j
+						outputIndexes = append(outputIndexes, outputNum)
+					}
+					unspentBits >>= 1
+				}
+				offset++
+			}
+
+			// Create a map to hold all of the converted outputs for the entry.
+			type convertedOut struct {
+				compressedAmount uint64
+				compressedScript []byte
+				scriptVersion    uint64
+			}
+			outputs := make(map[uint32]*convertedOut)
+
+			// Decode and add all of the outputs.
+			for _, outputIndex := range outputIndexes {
+				// Deserialize the compressed amount and ensure there are bytes
+				// remaining for the compressed script.
+				compressedAmount, bytesRead := deserializeVLQ(oldSerialized[offset:])
+				if bytesRead == 0 {
+					return errDeserialize("unexpected end of data during decoding " +
+						"(compressed amount)")
+				}
+				offset += bytesRead
+
+				// Decode the script version.
+				scriptVersion, bytesRead := deserializeVLQ(oldSerialized[offset:])
+				if bytesRead == 0 {
+					return errDeserialize("unexpected end of data during decoding " +
+						"(script version)")
+				}
+				offset += bytesRead
+
+				// Decode the compressed script size and ensure there are enough bytes
+				// left in the slice for it.
+				const compressionVersion = 1
+				scriptSize := decodeCompressedScriptSize(oldSerialized[offset:],
+					compressionVersion)
+				// Note: scriptSize == 0 is OK (an empty compressed script is valid)
+				if scriptSize < 0 {
+					return errDeserialize("negative script size")
+				}
+				if len(oldSerialized[offset:]) < scriptSize {
+					return errDeserialize(fmt.Sprintf("unexpected end of "+
+						"data after script size (got %v, need %v)",
+						len(oldSerialized[offset:]), scriptSize))
+				}
+				compressedScript := make([]byte, scriptSize)
+				copy(compressedScript, oldSerialized[offset:offset+scriptSize])
+				offset += scriptSize
+
+				// Create a converted utxo entry with the details deserialized above.
+				outputs[outputIndex] = &convertedOut{
+					compressedAmount: compressedAmount,
+					compressedScript: compressedScript,
+					scriptVersion:    scriptVersion,
+				}
+			}
+
+			// Copy the stake extra data if this was a ticket.
+			var ticketMinOuts []byte
+			if txType == stake.TxTypeSStx {
+				ticketMinOuts = make([]byte, len(oldSerialized[offset:]))
+				copy(ticketMinOuts, oldSerialized[offset:])
+			}
+
+			// Create V3 utxo entries with the details deserialized above.
+			//
+			// The V3 serialized key format is:
+			//
+			//   <hash><output index><tree>
+			//
+			//   Field                Type             Size
+			//   hash                 chainhash.Hash   chainhash.HashSize
+			//   output index         VLQ              variable
+			//   tree                 VLQ              variable
```

Comment:
Do you think it would be worth making the order `<hash><tree><output index>` instead so it's possible to iterate over a specific tree by creating a cursor with the prefix `<hash><tree>` and also the natural iteration order would then be sorted by tree, then block index, too?

It might come in handy for the cache and it seems like it would more logically match how they're actually structured within a block and thus likely to be looked up, perhaps allowing for faster sequential seeking versus the current layout that will require randomly jumping back and forth as it first looks up outputs from the regular tree and then the stake tree.

e.g. Currently:

```
<hash a><block index 0><tree 0>  <-- Loaded 1st
<hash a><block index 0><tree 1>  <-- Loaded 4th
<hash a><block index 1><tree 0>  <-- Loaded 2nd
<hash a><block index 1><tree 1>  <-- Loaded 5th
<hash a><block index 2><tree 0>  <-- Loaded 3rd
<hash a><block index 2><tree 1>  <-- Loaded 6th
```

Versus:
```
<hash a><tree 0><block index 0>  <-- Loaded 1st
<hash a><tree 0><block index 1>  <-- Loaded 2nd
<hash a><tree 0><block index 2>  <-- Loaded 3rd
<hash a><tree 1><block index 0>  <-- Loaded 4th
<hash a><tree 1><block index 1>  <-- Loaded 5th
<hash a><tree 1><block index 2>  <-- Loaded 6th
```

I realize this would apply elsewhere to the new v3 format and not just here.

This is definitely not something that is necessary to land this PR at all, just a thought.

Reply:
I thought about this a bit and went back and forth on it (also, it should be `output index` in your comment rather than `block index`).  On the one hand, I don't think it really matters either way, since:

- In practice, having the same hash for both trees would be extremely rare, so jumping back and forth to check both trees would never really happen.
- We want to avoid seeking in general, and we've removed all cases that required this up to this point and should be able to lookup by a proper outpoint key.

That being said, I decided to switch it to `<hash><tree><index>` for the following reasons:

- In the case where we know the `hash` but don't know the `tree` and/or `index`, I think it's preferred to have the `tree` come first.
    - In the case you have the `tree` but don't have the `index`:
        - You can easily scan `<hash><tree>...` to find the first index or all indexes, depending on what you are looking for.
    - In the case that you have the `index`, but don't have the `tree`:
        - You would need to check both trees, but since there are only 2 trees, this is still efficient.
- If we have the `index` come first, the main downside is the case where you know the `tree` but do not know the `index`:
    - Scanning would be a bit less straightforward in this case, as you'd have to scan `<hash><each output>` to find the first output, but then check the `tree` when an entry is found, and continue scanning if the `tree` doesn't match (and even though it would be extremely unlikely that the tree wouldn't match, we'd still have to check and handle that case).
- I do think that the structure makes more sense the `hash` and `tree` are used to identify the tx, whereas the index is used to identify the output, e.g.:
    ```
    |-tx unique identifier-| |---- tx output ----|
          <hash><tree>           <output index>
    ```

Since it's a bit cleaner and a bit more future proof (and was a small update), I went ahead and made that change.

Reply:
Great.  Indeed I originally said output index and then put block index in the following examples.  Oops.

I agree that we typically shouldn't generally be scanning/seeking anymore since we should be doing all lookups by the full outpoint.  However, I also know that sometimes things come up that require limited scanning and it's much harder to change keys in the database later.

As you note, the number of trees is significantly less than the output indexes (2 versus potentially 2^32-1) , so it significantly limits the search space in the event you don't have the full outpoint for whatever reason.

Come to think of it, I think we already have one such case when trying to determine if inventory is already known by falling back to the utxoset which ends up looking into both trees for a limited number of outputs.  It's not something to worry about for this PR, but that could probably be optimized now due to this change to do a cursor lookup of `<hash><tree>` for the two trees instead to determine if there are _any_ outputs.